Remove deprecated postgresql ensurePermissions

laalsaas opened the pull request:
Replaces deprecated option `postgresql.ensureUsers.*.ensurePermissions` with a custom systemd unit that ensures database privileges for k3s. This was the only place where we used the option.

PL-132355

@flyingcircusio/release-managers

## Release process

Impact:

Changelog:

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - This test should not change the behaviour compared to previous releases.
- [x] Security requirements tested? (EVIDENCE)
  - I locally changed the postgress-version for k3s to 16 and ran the test cases. All tests passed.

leona-ya commented on nixos/roles/k3s/server.nix:
```diff
@@ -304,12 +304,32 @@ in {
       } ];
     };
 
+    systemd.services.kubernetes-ensure-db-permissions = {
```
Is there a reason why to use a custom systemd service instead of `services.postgresql.ensureUsers.*.ensureDBOwnership`?

EDIT: I see. The username is differently. Still the other comments apply. This might be something we want to have in upstream nixos, but might also not (this is a huge discussion topic unfortunately).

leona-ya commented on nixos/roles/k3s/server.nix:
```diff
@@ -304,12 +304,32 @@ in {
+            exit 1
+        fi
+        done
+          $PSQL -tAc 'GRANT ALL PRIVILEGES ON DATABASE kubernetes TO "root"'
```
This will most likely also not work on PostgreSQL 15+. The user needs to have permissions on the `public` schema too or ownership of the DB. See https://github.com/NixOS/nixpkgs/issues/216989 for more information.

leona-ya commented on nixos/roles/k3s/server.nix:
```diff
@@ -304,12 +304,32 @@ in {
 
+    systemd.services.kubernetes-ensure-db-permissions = {
+      description = "Ensure the root user has all permissions on kubernetes db";
+      wantedBy = [ "k3s.service" ];
```
I think this should also have a 
```nix
Before = [ "k3s.service" ]
```

laalsaas committed:
Remove deprecated postgresql ensurePermissions

Replaces deprecated option `postgresql.ensureUsers.*.ensurePermissions`
with a custum systemd unit that ensures the privileges

PL-132355

laalsaas committed:
k3s: grant required permissions

This grants the root user the required CREATE privilege on the public
schema.

PL-132355

dpausp commented on nixos/roles/k3s/server.nix:
```diff
@@ -304,12 +304,33 @@ in {
       } ];
     };
 
+    systemd.services.kubernetes-ensure-db-permissions = {
```
better follow the naming scheme for existing services: `fc-k3s-ensure-db-permissions`
